File: cpp/vm.cpp
#include "utility.h"
#include "vm.h"


Code newPush(int i) {
    return Code(OP_PUSH, i);
}

Code newPlus() {
    return Code(OP_PLUS);
}

Code newMult() {
    return Code(OP_MULT);
}


Optional<int> VM::run() {
    // always start with an empty stack
    std::stack<int> d;
    s.swap(d);

    for(int i = 0; i < code.size(); i++) {
        switch(code[i].kind) {
        case OP_PUSH:
            s.push(code[i].val);
            break;
        case OP_MULT: {
            int right = s.top();
            s.pop();
            int left = s.top();
            s.pop();
            s.push(left * right);
            break;
        }
        case OP_PLUS: {
            int right = s.top();
            s.pop();
            int left = s.top();
            s.pop();
            s.push(left + right);
            break;
        }
        }
    }

    if (s.empty()) {
        return nothing<int>();
    }

    return just<int>(s.top());
} // run

// menschenlesbare Darstellung von VM-Instruktion
std::string Code::toString() const {
    switch (kind) {
    case OP_PUSH:
        return "Push " + std::to_string(val) + ";";
    case OP_PLUS:
        return "Plus;";
    case OP_MULT:
        return "Mult;";
    default:
        return "UNK;";
    }
}

File: rust/src/vm.rs
use core::fmt;

// Stack-based VM
// Supported instructions:
// Push <int>
// Plus
// Mult

pub enum Code {
    Push(i32),
    Plus,
    Mult,
}

pub struct VM {
    code: Vec<Code>,
    stack: Vec<i32>,
}

impl VM {
    pub fn new(code: Vec<Code>) -> Self {
        Self {
            code,
            stack: Vec::new(), // NOTE: rust automatically infers concrete type (Vec<i32>)
        }
    }

    // Result instead of Option to clearly communicate error
    pub fn run(&mut self) -> Result<i32, String> {
        // always start with empty stack
        self.stack.clear();

        // NOTE: iter -> borrow elements of vec (compare into_vec, which consumes vec)
        // enumerate to provide line number in errors
        for code in self.code.iter().enumerate() {
            match code.1 {
                Code::Push(val) => {
                    self.stack.push(*val);
                }
                Code::Plus => {
                    if self.stack.len() < 2 {
                        return Err(format!(
                            "L{} PLUS: not enough values on stack. Expected two",
                            code.0
                        ));
                    }
                    let right = self.stack.pop().unwrap();
                    let left = self.stack.pop().unwrap();
                    self.stack.push(left + right);
                }
                Code::Mult => {
                    if self.stack.len() < 2 {
                        return Err(format!(
                            "L{} MULT: not enough values on stack. Expected two",
                            code.0
                        ));
                    }
                    let right = self.stack.pop().unwrap();
                    let left = self.stack.pop().unwrap();
                    self.stack.push(left * right);
                }
            }
        }

        if let Some(result) = self.stack.last() {
            Ok(*result)
        } else {
            Err(format!(
                "L{} No return value found on stack",
                self.code.len()
            ))
        }
    }
}

// VM state to_string
impl fmt::Display for VM {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let code_string = self
            .code
            .iter()
            .map(|c| {
                match c {
                    Code::Push(val) => format!("Push {}", val),
                    Code::Plus => "Plus".to_string(),
                    Code::Mult => "Mult".to_string(),
                }
            })
            .collect::<Vec<String>>()
            .join(";");

        let stack_string = self
            .stack
            .iter()
            .map(|i| i.to_string())
            .collect::<Vec<String>>()
            .join(",");

        write!(f, "Code: {}", code_string)?;
        write!(f, "Stack: {}", stack_string)?;

        Ok(())
    }
}
